pkg/ingester/transfer.go

```diff
@@ -275,6 +402,67 @@ func (i *Ingester) transferOut(ctx context.Context) error {
 	return nil
 }
 
+func (i *Ingester) v2TransferOut(ctx context.Context) error {
+	if len(i.V2.dbs) == 0 {
+		level.Info(util.Logger).Log("msg", "nothing to transfer")
+		return nil
+	}
+
+	// Close all user databases
+	wg := &sync.WaitGroup{}
+	// Only perform a shutdown once
+	once.Do(func() {
+		wg.Add(len(i.V2.dbs))
+		for _, db := range i.V2.dbs {
+			go func(closer io.Closer) {
+				defer wg.Done()
+				if err := closer.Close(); err != nil {
+					level.Warn(util.Logger).Log("msg", "failed to close db", "err", err)
+				}
+			}(db)
+		}
+	})
+
+	targetIngester, err := i.findTargetIngester(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot find ingester to transfer blocks to: %v", err)
+	}
+
+	level.Info(util.Logger).Log("msg", "sending blocks", "to_ingester", targetIngester.Addr)
+	c, err := i.cfg.ingesterClientFactory(targetIngester.Addr, i.clientConfig)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	ctx = user.InjectOrgID(ctx, "-1")
+	stream, err := c.TransferTSDB(ctx)
+	if err != nil {
+		return errors.Wrap(err, "TransferTSDB")
+	}
+
+	wg.Wait() // wait for all databases to have closed
+
+	// Grab a list of all blocks that need to be shipped
+	blocks, err := unshippedBlocks(i.V2.TSDBDir)
+	if err != nil {
+		return err
+	}
+
+	for user, blockIDs := range blocks {
+		// Transfer the users TSDB
+		// TODO(thor) transferring users can be done concurrently
```

Comment:
In `transferUser()` we read the entire block files in memory and then send them in batches of 1MB each. If we concurrently send multiple tenants, we may go OOM. Would be glad to add this note to the comment for the future, so that we optimize the `batchSend()` before concurrently transferring users blocks.